lib/AST/ASTMangler.cpp

```diff
@@ -1470,48 +1469,76 @@ void ASTMangler::appendFunctionType(AnyFunctionType *fn,
   }
 }
 
-void ASTMangler::appendFunctionSignature(AnyFunctionType *fn,
-                                         bool forceSingleParam) {
-  appendParams(fn->getResult(), /*forceSingleParam*/ false);
-  appendParams(fn->getInput(), forceSingleParam);
+void ASTMangler::appendFunctionSignature(AnyFunctionType *fn) {
+  appendFunctionResultType(fn->getResult());
+  appendFunctionInputType(fn->getParams());
   if (fn->throws())
     appendOperator("K");
 }
 
-void ASTMangler::appendParams(Type ParamsTy, bool forceSingleParam) {
-  if (TupleType *Tuple = ParamsTy->getAs<TupleType>()) {
-    if (Tuple->getNumElements() == 0) {
-      if (forceSingleParam) {
-        // A tuple containing a single empty tuple.
-        appendOperator("y");
-        appendOperator("t");
-        appendListSeparator();
-        appendOperator("t");
-      } else {
-        appendOperator("y");
-      }
-      return;
+void ASTMangler::appendFunctionInputType(
+    ArrayRef<AnyFunctionType::Param> params) {
+  auto getParamType = [](const AnyFunctionType::Param &param) -> Type {
+    auto type = param.getType();
+
+    // FIXME: Change mangling for variadic parameters so
+    //        the enclosing array type is not required.
```

Comment:
I'm not sure what this FIXME means, but in general we should finalize the mangling right now.

Reply:
That would be incorrect because checking label and the param type are two different things e.g. `((Int, Int))` is going to form a single parameter with (Int, Int) tuple inside, and checking for presence of label first makes sure that it wasn't a single element tuple with tuple inside e.g. `(foo: (Int, Int))`, and then if the type of the element itself is a tuple it adds enclosing container to the name to make sure that we don't drop the sugar, otherwise we just encode it as a regular single type.

Reply:
I just use it too say that we need to address this separately.